pkg/generate/template.go

```diff
@@ -0,0 +1,67 @@
+package generate
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"regexp"
+	"strings"
+)
+
+var (
+	ErrCouldNotReadFile = errors.New("could not read file")
+)
+
+// Template contains an array of documents that can be
+// interacted with with its various functions.
+type template struct {
+	documents []string
+}
+
+func buildTemplate(file io.Reader) (template, error) {
+	// Separate the documents by the yaml separator and build a template with them
+	fileBytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		return template{}, ErrCouldNotReadFile
+	}
+
+	docs := strings.Split(string(fileBytes), "---")
+
+	constructedTemplate := template{}
+	for _, doc := range docs {
+		trimmedDoc := strings.TrimSpace(doc)
+		if trimmedDoc != "" {
+			constructedTemplate.documents = append(constructedTemplate.documents, trimmedDoc)
+		}
+	}
+
+	return constructedTemplate, nil
+}
+
+// has determines if any of the documents for the template
+// contains the specified string.
+func (t template) has(find string) bool {
+	for _, document := range t.documents {
+		if document == find {
+			return true
+		}
+	}
+	return false
+}
+
+// with builds the template documents with the combination set specified
+func (t template) with(combo map[string]string, to template) template {
```

Comment:
suggestion: since we're mutating `to` here would it make more sense to pass it in as a `*template` and remove the return value?

(not sure what the Go best practices would be here)

Reply:
This brings up a good point, I went ahead and refactored this to make a bit more sense given the mutation. This now works on the original template itself (optimizing memory a bit more) as well as being a bit easier to understand. The interface for interacting with this will now look like:

```go
file := strings.NewReader("hello: WORLD")

t, err := buildTemplate(file)
if err != nil {
    return err
}

t.with(combination)

fmt.Println(t.processedDocuments)
```
Check out [this commit](https://github.com/operator-framework/combo/pull/38/commits/c066182fbf665d1df3404f491ec2afc16ee8e8ce) for more details!

Thoughts?